crates/polars-io/src/file_cache/cache_lock.rs

```diff
@@ -0,0 +1,173 @@
+use std::path::PathBuf;
+use std::sync::atomic::AtomicBool;
+use std::sync::{Arc, RwLock, RwLockReadGuard, RwLockWriteGuard};
+use std::time::Duration;
+
+use fs4::FileExt;
+use once_cell::sync::Lazy;
+use polars_core::config;
+
+use crate::pl_async;
+
+pub(super) static GLOBAL_FILE_CACHE_LOCK: Lazy<GlobalLock> = Lazy::new(|| {
+    let path = std::env::var("POLARS_TEMP_DIR")
+        .unwrap_or_else(|_| std::env::temp_dir().to_string_lossy().into_owned());
+    let path = PathBuf::from(path).join("polars/file-cache/");
+    let _ = std::fs::create_dir_all(&path);
+    let path = path.join(".process-lock");
+
+    let file = std::fs::OpenOptions::new()
+        .write(true)
+        .create(true)
+        .truncate(false)
+        .open(path)
+        .map_err(|err| {
+            panic!("failed to open/create global file cache lockfile: {}", err);
+        })
+        .unwrap();
+
+    let at_bool = Arc::new(AtomicBool::new(false));
+    // Holding this access tracker prevents the background task from
+    // unlocking the lock.
+    let access_tracker = AccessTracker(at_bool.clone());
+
+    pl_async::get_runtime().spawn(async move {
+        let access_tracker = at_bool;
+        let verbose = config::verbose();
+
+        loop {
+            if !access_tracker.swap(false, std::sync::atomic::Ordering::Relaxed)
+                && GLOBAL_FILE_CACHE_LOCK.try_unlock()
+                && verbose
+            {
+                eprintln!("unlocked global file cache lockfile");
+            }
+            tokio::time::sleep(Duration::from_secs(3)).await;
+        }
+    });
+
+    GlobalLock {
+        inner: RwLock::new(GlobalLockData { file, state: None }),
+        access_tracker,
+    }
+});
+
+pub(super) enum LockedState {
+    Shared,
+    #[allow(dead_code)]
+    Exclusive,
+}
+
+#[allow(dead_code)]
+pub(super) type GlobalFileCacheGuardAny<'a> = RwLockReadGuard<'a, GlobalLockData>;
+pub(super) type GlobalFileCacheGuardExclusive<'a> = RwLockWriteGuard<'a, GlobalLockData>;
+
+#[derive(Clone)]
+struct AccessTracker(Arc<AtomicBool>);
```

Comment:
Can we add a comment explaining what this boolean means?

Reply:
This was a very good point - I've left a comment summarizing its purpose, but how it achieves its purpose would probably take a few paragraphs 😂